### DISCUSSION TITLE
Why can't I import CallAdapterProvider?

### DISCUSSION THREAD

**Discussion by strawhat-shima** [Azure/communication-ui-library#2381] [Q&A]
CallAdapter can be imported from @azure/communication-react, but CallAdapterProvider could not be imported. What is the reason?

**Comment by prabhjot-msft** [comment]
`CallAdapterProvider` is an internal provider. It is not part of the public API. Can you provide more context where do you need to use it?

**Reply by strawhat-shima** [reply]
Thank you for your reply. You're right, I understand. I would like to create MyCallComposite by referring to the source code of CallComposite. At that time, instead of using CallClientProvider/CallAgentProvider/CallProvider, shouldn't we use CallAdapterProvider? Thought it was. I'm puzzled by the unexpectedly large gap between using Composites or UIComponents.

In addition, ConfigurationPage, LocalPreview, LocalDeviceSettings... the list goes on and on, but I wish I could use it! There is something I think. I am mostly happy with CallComposite and only a few changes i want to implement in MyCallComposite;

**Reply by JamesBurnside** [reply]
Hi @strawhat-shima we are exploring how to reduce the gap between composites and ui components but unfortunately do not have an update we can provide you today. Currently the composites only support limited customizations; theming, custom data models, and localization. If you wish further customization you will need to create a custom experience by either forking this repo or using UI components. You can take a look at the source to recreate any component that is specific to composites or check out our [storybook](https://aka.ms/acsstorybook), and you can also use either the stateful client or the adapter interface to power these components. The adapter interface has methods like `adaper.onStateChange` that can be used to drive ui components using the state inside the adapter. But we understand this is a large task if you only want small customizations.

We would love to hear the customizations you would like to make but are unable to with the current Composites and see if these are features we could add at a later date.

**Reply by strawhat-shima** [reply]
Thank you for your reply. I am regard.
My boss has two requests.
・I want to store the comandbar after entering the room in the page header section
・I don't want my own tile in the video gallery to be small, so I want to be able to freely customize the gallery display method.
Thank you.

**Reply by JamesBurnside** [reply]
Thanks for the input. For your control bar you can take a look at hiding the default composite call control bar and creating a control bar of your own; I've created a [small gist](https://gist.github.com/JamesBurnside/4bd4ecbc4a7ab275abcf8f8dbc733c98) to highlight how this could be done. We'll look into what flexibility we can extend towards the VideoGallery layout from the composite interface however the Composites are designed to be a packaged entity that is feature rich but has limited customization so this may not become available. For very custom layouts we recommend creating your own experience using [UI Components](https://azure.github.io/communication-ui-library/?path=/docs/overview-uicomponents--page) and the [StatefulClients](https://azure.github.io/communication-ui-library/?path=/docs/statefulclient-overview--page).

**Reply by strawhat-shima** [reply]
Thank you for creating the code for reference. I don't understand it yet, but I will try.

**Comment by alkwa-msft** [comment]
Closing this discussion due to inactivity. Feel free to re-open if this is still an issue.
